File: src/main/java/com/dataArtisans/flink_training/exercises/datastream_java/popularPlaces/PopularPlaces.java
/*
 * Copyright 2015 data Artisans GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dataartisans.flink_training.exercises.datastream_java.popularplaces;

import com.dataartisans.flink_training.exercises.datastream_java.ridecleansing.RideCleansing;
import com.dataartisans.flink_training.exercises.datastream_java.sources.TaxiRideSource;
import com.dataartisans.flink_training.exercises.datastream_java.utils.GeoUtils;
import com.dataartisans.flink_training.exercises.datastream_java.datatypes.TaxiRide;
import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.FoldFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.api.java.tuple.Tuple4;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;

/**
 * Java reference implementation for the "Popular Places" exercise of the Flink training (http://dataartisans.github.io/flink-training).
 * The task of the exercise is to identify every five minutes popular areas where many taxi rides arrived or departed in the last 15 minutes.
 *
 * Parameters:
 *   -input path to input file
 *   -popThreshold minimum number of taxi rides for popular places
 *   -maxDelay maximum out of order delay of events
 *   -speed serving speed factor
 *
 */
public class PopularPlaces {

	public static void main(String[] args) throws Exception {

		// read parameters
		ParameterTool params = ParameterTool.fromArgs(args);
		String input = params.getRequired("input");
		final int popThreshold = Integer.parseInt(params.getRequired("popThreshold"));
		final int maxEventDelay = params.getInt("maxDelay", 0);
		final float servingSpeedFactor = params.getFloat("speed", 1.0f);

		// set up streaming execution environment
		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

		// start the data generator
		DataStream<TaxiRide> rides = env.addSource(
				new TaxiRideSource(input, maxEventDelay, servingSpeedFactor));

		// find n most popular spots
		DataStream<Tuple4<Float, Float, Boolean, Integer>> popularSpots = rides
				// remove all rides which are not within NYC
				.filter(new RideCleansing.NYCFilter())
				// match ride to grid cell and event type (start or end)
				.map(new GridCellMatcher())
				// partition by cell id and event type
				.keyBy(0, 1)
				// build sliding window
				.timeWindow(Time.minutes(15), Time.minutes(5))
				// count events in window
				.fold(new Tuple3<Integer, Boolean, Integer>(0, false, 0),
						new FoldFunction<Tuple2<Integer, Boolean>, Tuple3<Integer, Boolean, Integer>>() {
					@Override
					public Tuple3<Integer, Boolean, Integer> fold(
							Tuple3<Integer, Boolean, Integer> cnt,
							Tuple2<Integer, Boolean> ride) throws Exception {

						cnt.f0 = ride.f0;
						cnt.f1 = ride.f1;
						cnt.f2++;
						return cnt;
					}
				})
				// filter by pop threshold
				.filter(new FilterFunction<Tuple3<Integer, Boolean, Integer>>() {
					@Override
					public boolean filter(Tuple3<Integer, Boolean, Integer> count) throws Exception {
						return count.f2 >= popThreshold;
					}
				})
				// map grid cell to coordinates
				.map(new GridToCoordinates());

		// print result on stdout
		popularSpots.print();

		// execute the transformation pipeline
		env.execute("Popular Places");
	}

	/**
	 * Map taxi ride to grid cell and event type.
	 * Start records use departure location, end record use arrival location.
	 */
	public static class GridCellMatcher implements MapFunction<TaxiRide, Tuple2<Integer, Boolean>> {

		@Override
		public Tuple2<Integer, Boolean> map(TaxiRide taxiRide) throws Exception {
			if(taxiRide.isStart) {
				// get grid cell id for start location
				int gridId = GeoUtils.mapToGridCell(taxiRide.startLon, taxiRide.startLat);
				return new Tuple2<Integer, Boolean>(gridId, true);
			} else {
				// get grid cell id for end location
				int gridId = GeoUtils.mapToGridCell(taxiRide.endLon, taxiRide.endLat);
				return new Tuple2<Integer, Boolean>(gridId, false);
			}
		}
	}

	/**
	 * Maps the grid cell id back to longitude and latitude coordinates.
	 */
	public static class GridToCoordinates implements
			MapFunction<Tuple3<Integer, Boolean, Integer>, Tuple4<Float, Float, Boolean, Integer>> {

		@Override
		public Tuple4<Float, Float, Boolean, Integer> map(Tuple3<Integer, Boolean, Integer> cellCount) throws Exception {

			return new Tuple4<Float, Float, Boolean, Integer>(
					GeoUtils.getGridCellCenterLon(cellCount.f0),
					GeoUtils.getGridCellCenterLat(cellCount.f0),
					cellCount.f1,
					cellCount.f2);
		}
	}

}

File: src/main/scala/com/dataArtisans/flinkTraining/exercises/dataStreamScala/popularPlaces/PopularPlaces.scala
/*
 * Copyright 2015 data Artisans GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dataArtisans.flinkTraining.exercises.dataStreamScala.popularPlaces

import com.dataArtisans.flinkTraining.exercises.dataStreamJava.dataTypes.TaxiRide
import com.dataArtisans.flinkTraining.exercises.dataStreamJava.sources.TaxiRideSource
import com.dataArtisans.flinkTraining.exercises.dataStreamJava.utils.GeoUtils
import org.apache.flink.api.common.functions.MapFunction
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.scala._

/**
 * Scala reference implementation for the "Popular Places" exercise of the Flink training (http://dataartisans.github.io/flink-training).
 * The task of the exercise is to identify every five minutes popular areas where many taxi rides arrived or departed in the last 15 minutes.
 *
 * Parameters:
 * --input path-to-input-directory
 * -- popThreshold min-num-of-taxis-for-popular-places
 * --speed serving-speed-of-generator
 *
 */
object PopularPlaces {
  private val COUNT_WINDOW_LENGTH = 15 * 60 * 1000L // 15 minutes in msecs
  private val COUNT_WINDOW_FREQUENCY = 5 * 60 * 1000L // 5 minutes in msecs

  def main(args: Array[String]) {

    // read parameters
    val params = ParameterTool.fromArgs(args)
    val input = params.getRequired("input")
    val popThreashold = params.getInt("popThreshold")
    val servingSpeedFactor = params.getFloat("speed", 1.0f)

    // adjust window size and eviction interval to fast-forward factor
    val windowSize = (COUNT_WINDOW_LENGTH / servingSpeedFactor).toLong
    val evictionInterval = (COUNT_WINDOW_FREQUENCY / servingSpeedFactor).toLong

    // set up streaming execution environment
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    // start the data generator
    val rides = env.addSource(new TaxiRideSource(input, servingSpeedFactor))

//    // find n most popular spots
//    val popularSpots = rides
//      // remove all rides which are not within NYC
//      .filter { r => GeoUtils.isInNYC(r.startLon, r.startLat) && GeoUtils.isInNYC(r.endLon, r.endLat) }
//      // match ride to grid cell and event type (start or end)
//      .map(new GridCellMatcher)
//      // partition by cell id and event type
//      .groupBy(0, 1)
//      // build sliding window
//      .window(Time.of(windowSize, TimeUnit.MILLISECONDS))
//        .every(Time.of(evictionInterval, TimeUnit.MILLISECONDS))
//      // count events in window
//      .mapWindow(new PopularityCounter(popThreashold))
//      .flatten()
//      // map grid cell to coordinates
//      .map(new GridToCoordinates)
//
//    // print result on stdout
//    popularSpots.print()

    // execute the transformation pipeline
    env.execute("Popular Places")
  }

  /**
   * Map taxi ride to grid cell and event type.
   * Start records use departure location, end record use arrival location.
   */
  class GridCellMatcher extends MapFunction[TaxiRide, (Int, Boolean)] {

    def map(taxiRide: TaxiRide): (Int, Boolean) = {
      if (taxiRide.isStart) {
        // get grid cell id for start location
        val gridId: Int = GeoUtils.mapToGridCell(taxiRide.startLon, taxiRide.startLat)
        (gridId, true)
      } else {
        // get grid cell id for end location
        val gridId: Int = GeoUtils.mapToGridCell(taxiRide.endLon, taxiRide.endLat)
        (gridId, false)
      }
    }
  }

  /**
   * Count window events for grid cell and event type.
   * Only emits records if the count is equal or larger than the popularity threshold.
   */
//  class PopularityCounter(popThreshold: Int) extends WindowMapFunction[(Int, Boolean), (Int, Boolean, Int)] {
//
//    def mapWindow(values: java.lang.Iterable[(Int, Boolean)], out: Collector[(Int, Boolean, Int)]) {
//
//      // count records in window and build output record
//      val result = values.asScala.foldLeft((0, false, 0)) { (l, r) => (r._1, r._2, l._3 + 1) }
//
//      // check threshold
//      if (result._3 > popThreshold) {
//        // emit record
//        out.collect(result)
//      }
//    }
//  }

  /**
   * Maps the grid cell id back to longitude and latitude coordinates.
   */
  class GridToCoordinates extends MapFunction[(Int, Boolean, Int), (Float, Float, Boolean, Int)] {

    def map(cellCount: (Int, Boolean, Int)): (Float, Float, Boolean, Int) = {
      val longitude = GeoUtils.getGridCellCenterLon(cellCount._1)
      val latitude = GeoUtils.getGridCellCenterLat(cellCount._1)
      (longitude, latitude, cellCount._2, cellCount._3)
    }
  }

}

